src/main/java/org/broadinstitute/hellbender/tools/spark/RevertSamSpark.java

```diff
@@ -0,0 +1,706 @@
+        final FeatureReader<TableFeature>  parser = AbstractFeatureReader.getFeatureReader(outputMapFile.getAbsolutePath(), new TableCodec(null), false);
+        for (final TableFeature row : parser.iterator()) {
+            final String id = row.get("READ_GROUP_ID");
+            final String output = row.get("OUTPUT");
+            final File outputPath = new File(output);
+            outputMap.put(id, outputPath);
+        }
+        CloserUtil.close(parser);
+        return outputMap;
+    }
+
+    // Names the files based on the readgroups individually presented in the header
+    private static Map<String, File> createOutputMapFromHeader(final List<SAMReadGroupRecord> readGroups, final File outputDir, final String extension) {
+        final Map<String, File> outputMap = new HashMap<>();
+        for (final SAMReadGroupRecord readGroup : readGroups) {
+            final String id = readGroup.getId();
+            final String fileName = id + extension;
+            final Path outputPath = Paths.get(outputDir.toString(), fileName);
+            outputMap.put(id, outputPath.toFile());
+        }
+        return outputMap;
+    }
+
+// ________________________________________________________________________________________________________________________
+// sum garbage
+// ________________________________________________________________________________________________________________________
+    /**
+     * Methods used for validating parameters to RevertSam.
+     */
+    static class ValidationUtil {
+
+        static void validateOutputParams(final boolean outputByReadGroup, final File output, final File outputMap, final List<String> errors) {
+            if (outputByReadGroup) {
+                validateOutputParamsByReadGroup(output, outputMap, errors);
+            } else {
+                validateOutputParamsNotByReadGroup(output, outputMap, errors);
+            }
+        }
+
+        static void validateOutputParamsByReadGroup(final File output, final File outputMap, final List<String> errors) {
+            if (output != null) {
+                if (!Files.isDirectory(output.toPath())) {
+                    errors.add("When '--output-by-readgroup' is set and output is provided, it must be a directory: " + output);
+                }
+                return;
+            }
+            // output is null if we reached here
+            if (outputMap == null) {
+                errors.add("Must provide either output or outputMap when '--output-by-readgroup' is set.");
+                return;
+            }
+            if (!Files.isReadable(outputMap.toPath())) {
+                errors.add("Cannot read outputMap " + outputMap);
+                return;
+            }
+            final FeatureReader<TableFeature>  parser = AbstractFeatureReader.getFeatureReader(outputMap.getAbsolutePath(), new TableCodec(null),false);
+            if (!RevertSamSpark.ValidationUtil.isOutputMapHeaderValid((List<String>)parser.getHeader())) {
```

Comment:
you need to add a `SuppressWarnings("unchecked")` to this method or you'll get a warning.